### ISSUE TITLE
Order gets out-of-sync with parallel updates

### ISSUE THREAD

**Issue by johnnyshields** [mongoid/mongoid_orderable#44]
I've had a nagging issue with Mongoid::Orderable where, over time, randomly in my production database the `position` field will start looking like:

```
0, 0, 1, 1, 2, 3, 4, 4, 4, etc.
```

My theory as to why this happens is:

``` ruby
        def apply_position column, target_position
           ...
          elsif target_position < pos
            MongoidOrderable.inc(scope.where(col.gte => target_position, col.lt => pos), col, 1)
          elsif target_position > pos
            MongoidOrderable.inc(scope.where(col.gt => pos, col.lte => target_position), col, -1)
          end
        end
```

The `apply_position` method increments the range between "current_position" and "target_position". If someone is moving lots of items around quickly, I believe we can get a stale value for the current_position--possibly from a replica set instance, or maybe a race condition between sending two "move_to" requests to two separate web servers. In other words, in the current implementation we can't be 100% sure that the current_position is accurate.

For now, I'm going to work around this issue by adding a delay on the frontend, so when you move one object you can't trigger another request under the server responds. It would be better if we had an atomic way to read and set the position.

**Comment by johnnyshields** [comment]
@pyromaniac any thoughts on this?

**Comment by pyromaniac** [comment]
I didn't work with mongo too long, there should be a way to perform such an atomic update right now, so if you will provide a patch - it would be just awesome

**Comment by dblock** [comment]
I think we've seen instances of this issue. I am not saying I will try, but if someone is motivated to reproduce it, http://artsy.github.io/blog/2013/02/09/data-corruption-and-concurrent-updates-to-embedded-objects-with-mongoid/ could be helpful.

**Comment by dblock** [comment]
Turned out to be easier to reproduce than I thought ;(

https://github.com/pyromaniac/mongoid_orderable/pull/47

**Comment by samuelebistoletti** [comment]
Does anyone tried to fix that?

**Comment by dblock** [comment]
Doesn't look so @samuelebistoletti.

**Comment by johnnyshields** [comment]
@dblock very happy that I'm not alone in experiencing this. Any idea for possible fixes?

**Comment by dblock** [comment]
By taking a lock during update, either by using or by copying some code out of https://github.com/mongoid/mongoid-locker.

**Comment by johnnyshields** [comment]
I have a fix for this using transactions. Will raise PR soon.

**Comment by johnnyshields** [comment]
Will be fixed in version 6.0.0
